scalanet/ut/src/io/iohk/scalanet/peergroup/DynamicTLSPeerGroupSpec.scala

```diff
@@ -349,6 +375,142 @@ class DynamicTLSPeerGroupSpec extends AsyncFlatSpec with BeforeAndAfterAll {
+        _ <- Task(srvChannelImpl.incomingQueueSize).restartUntil(qSize => qSize == firstBatch.size)
+        _ <- Task.sleep(1.seconds)
+        // last two messages will be stored in rcv buffer, becouse incoming queue is full
+        _ <- Task.traverse(secondBatch)(m => clientChannel.sendMessage(m))
+        _ <- Task.sleep(1.seconds)
+        _ <- Task(assert(srvChannelImpl.incomingQueueSize == firstBatch.size))
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(firstBatch(0))))
+        _ <- Task(assert(srvChannelImpl.incomingQueueSize == firstBatch.size - 1))
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(firstBatch(1))))
+        // after reading first 2 messages, there place in queue for two new messages, so they are read from rcv buffer
+        _ <- Task(srvChannelImpl.incomingQueueSize)
+          .restartUntil(qSize => qSize == ((firstBatch.size - 2) + secondBatch.size))
+        _ <- Task(assert(srvChannelImpl.incomingQueueSize == 4))
+        // all received messages are in expected order
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(firstBatch(2))))
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(firstBatch(3))))
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(secondBatch(0))))
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(secondBatch(1))))
+        finalAssert <- Task(assert(srvChannelImpl.incomingQueueSize == 0))
+      } yield {
+        finalAssert
+      }
+  }
+
+  it should "communicate with slower consumer with small queue available" in backPressureTestCase(
+    clientConfig = getCorrectConfig(maxFrameLength = 64 * 1024 * 1024),
+    serverConfig = getCorrectConfig(maxQueueSize = 1, maxFrameLength = 64 * 1024 * 1024)
+  ) {
+    case (clientChannel, serverChannel) =>
+      val srvChannelImpl =
+        serverChannel.asInstanceOf[DynamicTLSPeerGroupInternals.DynamicTlsChannel[ChannelEvent[SizeAbleMessage]]]
+      // large message so fill the buffer after each sent
+      val messagesToSend = (0 to 24).map(_ => SizeAbleMessage.genRandomOfsizeN(4 * 1024 * 1024))
+      val sendTask = Task.traverse(messagesToSend)(m => clientChannel.sendMessage(m))
+      val receiveTask = serverChannel.nextChannelEvent
+        .delayExecution(500.milliseconds)
+        .toObservable
+        .collect { case MessageReceived(m) => m }
+        .take(messagesToSend.size)
+        .toListL
+      for {
+        receivedMessages <- Task.parMap2(sendTask, receiveTask)((_, received) => received)
+        _ <- Task(assert(messagesToSend == receivedMessages))
+        finalQueueSize <- Task(
+          serverChannel.asInstanceOf[DynamicTLSPeerGroupInternals.DynamicTlsChannel[_]].incomingQueueSize
+        )
+      } yield {
+        assert(finalQueueSize == 0)
+      }
```

Comment:
What is this testing exactly? It looks like it's checking that the receiver side has consumed all messages from the queue. Since you explicitly `.take(messagesToSend.size)` items from it, what else would the final count be other than 0?

Reply:
It is checking that explicit queue size counter in `ChannelAwareQueue` is decremented and incremented correctly.  If we would forgot to decrement it after consuming message then this test would fail